src/main/java/org/openhab/tools/analysis/checkstyle/EshInfXmlCheck.java

```diff
@@ -0,0 +1,348 @@
+
+        // Check for missing referenced config descriptions
+        Map<String, File> missingConfigDescriptions = intersect(allConfigDescriptionRefs, allConfigDescriptions);
+        logMissingEntries(missingConfigDescriptions, MESSAGE_MISSING_URI_CONFIGURATION);
+
+        // Check for unused bridge-type-refs.
+        Map<String, File> unusedBridges = intersect(allBridgeTypes, allSupportedBridges);
+        logMissingEntries(unusedBridges, MESSAGE_UNUSED_BRIDGE);
+
+        // Check for unused referenced config descriptions
+        Map<String, File> unusedConfigDescriptions = intersect(allConfigDescriptions, allConfigDescriptionRefs);
+        logMissingEntries(unusedConfigDescriptions, MESSAGE_UNUSED_URI_CONFIGURATION);
+
+        // Check for missing ESH-INF files in the build.properties
+        checkBuildProperties();
+    }
+
+    private void checkBuildProperties() {
+        if (buildPropertiesFile != null) {
+            IBuildEntry binIncludes = buildPropertiesFile.getEntry(IBuildEntry.BIN_INCLUDES);
+            if (binIncludes != null) {
+                String[] includedTokens = binIncludes.getTokens();
+                // Exclude processed files that are included
+                for (String included : includedTokens) {
+                    // Iterator is used, as the collection will be modified
+                    for (Iterator<Entry<Path, File>> it = eshInfFiles.entrySet().iterator(); it.hasNext();) {
+                        Map.Entry<Path, File> entry = it.next();
+                        if (entry.getKey().startsWith(included)) {
+                            it.remove();
+                        }
+                    }
+                }
+            }
+            logMissingEntries(eshInfFiles, MESSAGE_NOT_INCLUDED_XML_FILE);
+        }
+    }
+
+    private void processXmlFile(File xmlFile) throws CheckstyleException {
+        if (isEmpty(xmlFile)) {
+            log(0, MessageFormat.format(MESSAGE_EMPTY_FILE, xmlFile.getName()), xmlFile.getPath());
+        } else {
+
+            File fileParentDirectory = xmlFile.getParentFile();
+            boolean isESHParentDirectory = ESH_INF_DIRECTORY.equals(fileParentDirectory.getParentFile().getName());
+
+            if (isESHParentDirectory) {
+                Path filePath = xmlFile.toPath();
+                Path bundlePath = filePath.getParent().getParent().getParent();
+                Path relativePath = bundlePath.relativize(filePath);
+                eshInfFiles.put(relativePath, xmlFile);
+
+                String schema = null;
+                switch (fileParentDirectory.getName()) {
+                    case THING_DIRECTORY: {
+                        schema = thingSchema;
+                        // Process the files for all nodes below,
+                        // the allowed values are described in the thing description XSD
+                        Map<String, File> supportedBridgeTypes = evaluateExpressionOnFile(xmlFile,
```

Comment:
Good point, but should this not be defined in the XSD files :

> Occurrence Indicators
Occurrence indicators are used to define how often an element can occur.
Note: For all "Order" and "Group" indicators (any, all, choice, sequence, group name, and group reference) the default value for maxOccurs and minOccurs is 1. 

See https://www.w3schools.com/xml/schema_complex_indicators.asp.

From what I have read I am left with the impression that we shouldn't write logic to check for duplicates and leave that to the XML validation. - http://stackoverflow.com/questions/26341791/duplicates-in-xml.